core/app/Utils.hs

```diff
@@ -37,27 +38,39 @@ defaultPrograms =
            , cabalProgram = "cabal"
            , stackProgram = "stack" }
 
-constructCells :: [String] -> Int -> [Cell] -> [Cell]
-constructCells [] _ curr = map (\c -> c { cellContent = T.strip $ cellContent c}) filtered
-  where filtered = filter (\c -> (T.stripStart $ cellContent c) /=  (T.pack "")) curr
-constructCells lst@(x:xs) i curr =
-  if isPrefixOf "--" x then
-    constructCells lst' (i+1) (curr ++ [textcell])
-  else
-    constructCells lst'' (i+1) (curr ++ [codecell])
+listofCellContentLines :: [String] -> [[String]]
+listofCellContentLines = split whenLineIsFromATextCell
   where
-    lst' = dropWhile (\x -> isPrefixOf "--" x) lst
-    textcell = Cell TextCell i (T.pack $ unlines $ map (\('-':'-':' ':xs) -> xs) $ takeWhile (\x -> isPrefixOf "--" x) lst)
-    lst'' = dropWhile (\x -> not $ isPrefixOf "--" x) lst
-    codecell = Cell CodeCell i (T.pack $ unlines $ takeWhile (\x -> not $ isPrefixOf "--" x) lst)
+    whenLineIsFromATextCell = dropBlanks . condense $ whenElt (isPrefixOf "--")
+
+indexedCellContents :: [[String]] -> [(Int, [String])]
+indexedCellContents cellContentLines = zip [0..length cellContentLines] cellContentLines
+
+buildCell :: (Int, [String]) -> Cell
+buildCell (cellId, cellContent)
+  | cellContent `areFrom` TextCell = Cell TextCell cellId $ processedContents cellContent
+  | cellContent`areFrom` CodeCell = Cell CodeCell cellId $ processedContents cellContent
+  where
+    areFrom cc TextCell  = any (isPrefixOf "--") cc
+    areFrom cc CodeCell  = any (not . isPrefixOf "--") cc
+    processedContents = T.pack . unlines . map removeCommentCharacters
+    removeCommentCharacters line = if "--" `isPrefixOf` line then drop 3 line else line
```

Comment:
Just realized that this would make a bug when decoding code cells, as it would remove indentation from their lines.